File: 2414100138-张秀华/第一周作业/2414100138-张秀华-第一周代码/app/src/main/java/com/example/music/MusicService.java
package com.example.music;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.content.res.AssetFileDescriptor;
import android.media.MediaPlayer;
import android.os.Build;
import android.os.IBinder;

import androidx.annotation.Nullable;
import androidx.core.app.NotificationCompat;

import java.io.IOException;

public class MusicService extends Service {
    private static final String CHANNERLID = "startMusic";
    private MediaPlayer mMediaPlayer = new MediaPlayer();
    private static final String ACTION_START = "START";
    private static final String ACTION_STOP = "STOP";

    @Override
    public void onCreate() {
        super.onCreate();
        initMediaPlayer();
    }

    //初始化MediaPlayer
    private void initMediaPlayer() {
        try {
            AssetFileDescriptor fd = getAssets().openFd("music.mp3");
            mMediaPlayer.setDataSource(fd.getFileDescriptor(), fd.getStartOffset(), fd.getLength());
            mMediaPlayer.prepare();
        } catch (IOException e) {
            e.printStackTrace();
        }

    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {


        //管理音乐播放
        String action = intent.getAction();
        if (action != null) {
            switch (action) {
                case ACTION_START:
                    if (mMediaPlayer != null && !mMediaPlayer.isPlaying()) {
                        mMediaPlayer.start();
                    }
                    break;
                case ACTION_STOP:
                    if (mMediaPlayer != null) {
                        mMediaPlayer.stop();
                        mMediaPlayer.release();
                        initMediaPlayer();
                    }
                    stopForeground(STOP_FOREGROUND_REMOVE);
                    stopSelf();
                    break;
            }
        }

        //创建通知+前台服务
        String content = intent.getStringExtra("content");

        NotificationManager manager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel channel = new NotificationChannel(CHANNERLID, "后台播放音乐", NotificationManager.IMPORTANCE_DEFAULT);
            manager.createNotificationChannel(channel);
        }

        Intent intentNotification = new Intent(this, MainActivity.class);
        PendingIntent pi = PendingIntent.getActivity(this, 0, intentNotification, PendingIntent.FLAG_IMMUTABLE);

        Notification notification = new NotificationCompat.Builder(this, CHANNERLID)
                .setContentTitle("音乐播放器")
                .setContentText(content)
                .setSmallIcon(R.drawable.baseline_music_note_24)
                .setContentIntent(pi)
                .build();

        manager.notify(1, notification);

        return super.onStartCommand(intent, flags, startId);
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (mMediaPlayer != null) {
            mMediaPlayer.release();
            mMediaPlayer = null;
        }
    }
}

File: 2414100138-张秀华/第十周作业/2414100138-张秀华-第十周代码/Music_Kotlin/app/src/main/java/com/example/music_kotlin/MusicService.kt
package com.example.music_kotlin

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Context
import android.content.Intent
import android.content.res.AssetFileDescriptor
import android.media.MediaPlayer
import android.os.Build
import android.os.IBinder
import androidx.core.app.NotificationCompat
import java.io.IOException


class MusicService : Service() {

    // Kotlin 推荐常量名用大写下划线或小驼峰，这里按规范改为小驼峰
    private val channelId = "startMusic"
    private var mMediaPlayer: MediaPlayer? = null
    private val actionStart = "START"
    private val actionStop = "STOP"

    override fun onCreate() {
        super.onCreate()
        initMediaPlayer()
    }

    private fun initMediaPlayer() {
        try {
            val fd: AssetFileDescriptor = assets.openFd("music.mp3")
            mMediaPlayer = MediaPlayer().apply {
                setDataSource(fd.fileDescriptor, fd.startOffset, fd.length)
                prepare()
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        val action = intent?.action
        if (action != null) {
            when (action) {
                actionStart -> {
                    mMediaPlayer?.let { player ->
                        if (!player.isPlaying) {
                            player.start()
                        }
                    }
                }
                actionStop -> {
                    mMediaPlayer?.let { player ->
                        player.stop()
                        player.release()
                    }
                    mMediaPlayer = null
                    initMediaPlayer()

                    stopForeground(STOP_FOREGROUND_REMOVE)
                    stopSelf()
                }
            }
        }

        // 通知创建
        val content = intent?.getStringExtra("content") ?: ""
        val manager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

        // 1. 创建通知渠道（API 26+ 必须）
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(
                channelId,
                "后台播放音乐",
                NotificationManager.IMPORTANCE_DEFAULT
            )
            manager.createNotificationChannel(channel)
        }

        // 2. 创建 PendingIntent
        val intentNotification = Intent(this, MainActivity::class.java)
        val pi = PendingIntent.getActivity(
            this,
            0,
            intentNotification,
            PendingIntent.FLAG_IMMUTABLE
        )

        // 3. 构建通知
        val notification: Notification = NotificationCompat.Builder(this, channelId)
            .setContentTitle("音乐播放器")
            .setContentText(content)
            .setSmallIcon(R.drawable.baseline_music_note_24)
            .setContentIntent(pi)
            .build()

        // 4. 发送通知
        manager.notify(1, notification)

        return super.onStartCommand(intent, flags, startId)
    }

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    override fun onDestroy() {
        super.onDestroy()
        mMediaPlayer?.release()
        mMediaPlayer = null
    }
}